src/psyclone/psyad/transformations/preprocess.py

```diff
@@ -98,6 +88,20 @@ def preprocess_trans(kernel_psyir, active_variable_names):
             # Apply MATMUL transformation
             matmul_trans.apply(call)
 
+    # Replace array-ranges with explicit loops
+    for assignment in kernel_psyir.walk(Assignment):
+        if node_is_passive(assignment, active_variable_names):
+            # No need to modify passive assignments
+            continue
+        try:
+            arrayassign2loops_trans.apply(assignment)
+        except TransformationError:
+            # Double-check that the transformation succeeded in
```

Comment:
That feels very hard to understand. After some digging around, I assume this happens if the assignment does not contain a range? Either it needs better documentation - atm it sounds like the transformation succeeded, but raised an exception(??) - while I think it should mean "if the transformation couldn't be applied because it wasn't an array range"?
Or perhaps a slight code change: do the walk first to verify if there is a range, then call the transformation?

Reply:
Oh yes, that's better. I did think about doing that but then decided not to for some reason. Anyway, it works nicely and is much clearer.

Reply:
I've tested this with a couple of the LFRic kernels that first flagged the problem and it seems OK. @DrTVockerodtMO are you in a position now to test this branch with the full adjoint test app?

Reply:
I still do not know how to set that up yet, but if it produced the same KGO as before then it should be okay.

Reply:
Much better, thanks.